ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -56,17 +82,65 @@ def obstacle_cb(self, msg):
         # TODO: Callback for /obstacle_waypoint message. We will implement it later
         pass
 
+    def send_waypoints(self):
+        #generates the list of LOOKAHEAD_WPS waypoints based on location of car
+        #for now assume loop
+        new_wps_list = []
+        start_wps_ptr = self.closest_waypoint()
+        end_wps_ptr = (start_wps_ptr + LOOKAHEAD_WPS) % len(self.waypoints)
+        rospy.loginfo("waypoint_updater:send_waypoints start_wps_ptr = %d, end_wps_ptr = %d",
+                      start_wps_ptr, end_wps_ptr)
+        if end_wps_ptr > start_wps_ptr:
+            for w_p in self.waypoints[start_wps_ptr:end_wps_ptr]:
+                w_p.twist.twist.linear.x = DEFAULT_VELOCITY #
+                new_wps_list.append(w_p)
+            # end of for
+        else:
+            for w_p in self.waypoints[start_wps_ptr:]:
+                w_p.twist.twist.linear.x = DEFAULT_VELOCITY #
+                new_wps_list.append(w_p)
+            # end of for                
+            for w_p in self.waypoints[:end_wps_ptr]:
+                w_p.twist.twist.linear.x = DEFAULT_VELOCITY #
+                new_wps_list.append(w_p)
+            # end of for
+        # end of if 
+        rospy.loginfo("waypoint_updater:send_waypoints final_waypoints list length is %d",
+            len(new_wps_list))
+        lane = Lane()
+        lane.waypoints = list(new_wps_list)
+        lane.header.frame_id = '/world'
+        lane.header.stamp = rospy.Time()
+        self.pubs['final_waypoints'].publish(lane)
+        #pass
+
+
     def get_waypoint_velocity(self, waypoint):
         return waypoint.twist.twist.linear.x
 
-    def set_waypoint_velocity(self, waypoints, waypoint, velocity):
-        waypoints[waypoint].twist.twist.linear.x = velocity
-
-    def distance(self, waypoints, wp1, wp2):
+    def set_waypoint_velocity(self, waypointlist, waypoint, velocity):
+        waypointlist[waypoint].twist.twist.linear.x = velocity
+
+    def closest_waypoint(self):
+        dist = 1000000  #maybe should use max
+        closest = 0
+        distance_lambda = lambda a, b: math.sqrt((a.x-b.x)**2 + (a.y-b.y)**2)
+        for i in range(len(self.waypoints)):
+            tmpdist = distance_lambda(self.waypoints[i].pose.pose.position, self.pose.position)
+            if tmpdist < dist:
+                closest = i
+                dist = tmpdist
+            # end of if
+        # end of for
+        return closest
```

Comment:
Ok for now, but I read somewhere that the Pure Pursuit algorithm might have strange output if the first waypoint is actually behind the vehicle so we may need to add logic later to guarantee the closest waypoint is ahead (like the angle check in the Path Planning project).

Reply:
I will mark this as #TODO in the code.